File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        Stack<Integer> st = new Stack<>();
        st.push(0);
        
        for(int i = 1; i < temperatures.length; ++i) {
            while(!st.empty() && temperatures[i] > temperatures[st.peek()]) {
                res[st.peek()] = i - st.peek();
                st.pop();
            }
            st.push(i);
        }
        
        return res;
        
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& ts) {
        vector<int> res(ts.size());
        stack<pair<int, int>> st;
        for(int i = 0; i < ts.size(); ++i) {
            if(st.empty() || ts[i] <= st.top().first) {
                st.push({ts[i], i});
                continue;
            }
            else {
                while(!st.empty() and ts[i] > st.top().first) {
                    auto p = st.top(); st.pop();
                    res[p.second] = i - p.second;
                }
                st.push({ts[i], i});
            }
        }
        
        return res;
    }
};